src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java

```diff
@@ -0,0 +1,364 @@
+			BasicFileAttributes attributes = Files.readAttributes(pathToFile, BasicFileAttributes.class);
+			
+			Instant fileCreateTime = attributes.creationTime().toInstant();
+			Instant now = Instant.now();
+			//If the file is older than the maxAge...
+			if (fileCreateTime.isBefore( now.minus(this.maxAge) ))
+			{
+				//TODO: Option to save back-ups of old files.
+				logger.debug("File {} is older than allowed amount ({}) so it will be downloaded again.",this.destination,this.maxAge);
+				downloadData();
+				logger.debug("Download is complete.");
+			}
+			else
+			{
+				logger.debug("File {} is not older than allowed amount ({}) so it will not be downloaded.",this.destination,this.maxAge);
+			}
+		}
+		else
+		{
+			logger.debug("File {} does not exist and must be downloaded.", pathToFile);
+			//if file does not exist, get it from the URL.
+			downloadData();
+		}
+		
+		// Print some basic file stats.
+		if (Files.exists(pathToFile))
+		{
+			if (Files.isReadable(pathToFile))
+			{
+				BasicFileAttributes attribs = Files.readAttributes(pathToFile, BasicFileAttributes.class);
+				logger.info("File Info: Name: {}, Size: {}, Created: {}, Modified: {}",this.destination,attribs.size(), attribs.creationTime(), attribs.lastModifiedTime());
+			}
+			else
+			{
+				logger.error("File {} is not readable!", pathToFile);
+			}
+		}
+		else
+		{
+			// If something failed during the data retrieval, the file might not exist. If that happens, let the user know.
+			logger.error("File \"{}\" still does not exist after executing the file retriever!", pathToFile);
+		}
+	}
+
+	protected void downloadData() throws Exception
+	{
+		logger.trace("Scheme is: {}", this.uri.getScheme());
+		Path path = null;
+		try
+		{
+			path = Paths.get(this.destination);
+			Files.createDirectories(path.getParent());
+			if (this.uri.getScheme().equals("http") || this.uri.getScheme().equals("https"))
+			{
+				
+				doHttpDownload(path);
+			}
+			else if (this.uri.getScheme().equals("ftp") || this.uri.getScheme().equals("sftp"))
+			{
+				doFtpDownload();
+			}
+			else
+			{
+				throw new UnsupportedSchemeException("URI "+this.uri.toString()+" uses an unsupported scheme: "+this.uri.getScheme());
+			}
+		}
+		catch (URISyntaxException e)
+		{
+			logger.error("Error creating for download destination: " + this.destination, e);
+			e.printStackTrace();
+		}
+		catch (IOException e)
+		{
+			logger.error("Unable to create parent directory of download destination: " + path.toString(), e);
+			e.printStackTrace();
+		}
+		catch (Exception e)
+		{
+			logger.error("Error performing download!", e);
+			throw e;
+		}
+		
+
+	}
+
+	protected void doFtpDownload() throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		doFtpDownload(null, null);
+	}
+	
+	protected void doFtpDownload(String user, String password) throws SocketException, IOException, FileNotFoundException, Exception
+	{
+		user = user == null || user.trim().equals("") ? "anonymous" : user;
```

Comment:
Maybe:

```
if (user == null || user.trim().equals("")) {
    user = "anonymous"
}
```
I find I need to do a double-take to realize it's checking and resetting itself if necessary.

Reply:
Ew. What was I thinking?? :facepalm: